### PR TITLE
Cleanup XPath.StringToDouble

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Split from #54103 . This is the simplest part.

Breaks bug-to-bug compliance with XML (in comment). I think it should be OK.

### CODE REVIEW

**Comment by krwq** [comment]
> Breaks bug-to-bug compliance with XML (in comment). I think it should be OK

@huoyaoyuan can you elaborate on this a bit?

As far as I know TryParse depends slightly on the current culture - I do not think XML depends on that though which means on some cultures you might get different result...

Can you point to a specific page/chapter in the spec related to this piece of code? We should understand all differences here

**Comment by huoyaoyuan** [comment]
> As far as I know TryParse depends slightly on the current culture

I have passed `CultureInfo.InvariantCulture`.

> Can you point to a specific page/chapter in the spec

https://www.w3.org/TR/xquery-operators/#numeric-types

The spec mentions that numbers are IEEE754.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPathConvert.cs [hunk 1]
```diff
@@ -2961,236 +2961,14 @@ public static string DoubleToString(double dbl)
...
-                            goto LGetRight;
-                        }
-                        break;
-                    default:
-                        // MSXML has a bug, we should not allow whitespace after a minus sign
-                        if (IsWhitespace(ch) && sign > 0)
-                        {
-                            pch = SkipWhitespace(pch);
-                            goto LRestart;
-                        }
```
Comment (huoyaoyuan) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPathConvert.cs:3041]: @krwq This is the bug. If there is whitespace after minus sign, the state of parser is reset in MSXML.

**Comment by huoyaoyuan** [comment]
https://www.w3.org/TR/xslt-30/#func-same-key This section indicates that infinities are parsed for 'INF' and '-INF', however the original code didn't handle this.

**Comment by jeffhandley** [comment]
@krwq This PR is assigned to you for follow-up/decision before the RC1 snap.

**Comment by krwq** [comment]
@huoyaoyuan @jeffhandley I'd prefer to wait with this until after RC1 snap - I'd prefer to have a better understanding on exact differences here and in case we miss something I'd rather have more time to react than bringing this last minute into LTS

**Comment by jeffhandley** [comment]
Sounds good to me, @krwq. To reflect that, I'm going to:

1. Set the milestone to 7.0.0
2. Apply the NO MERGE label
3. Convert this to a Draft

After main becomes 7.0, we can proceed merging it in.

**Comment by stephentoub** [comment]
@jeffhandley, did you intend for this to be closed? Making things drafts opens them up to being closed by the bot after a period of inactivity, in this case entirely because of us.

**Comment by krwq** [comment]
@huoyaoyuan I think the change looks good but I'd prefer we've added some tests here (or at least pointed to existing tests) and checked with old and new code and check some tricky inputs based on the old code

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPathConvert.cs (hunk 1 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPathConvert.cs:3041]: I think this change is ok, any other diffs you're aware of? I'd prefer we've added some tests here

**Comment by huoyaoyuan** [comment]
@krwq I'm not professional for XML/XSLT so I don't know how to construct the tests.

**Comment by krwq** [comment]
@huoyaoyuan example to add test is this project:
https://github.com/dotnet/runtime/tree/main/src/libraries/System.Private.Xml/tests/Misc

you'll need to write a tiny code which hits this code path and ideally:
- takes string as input
- indentify different edge cases
- compare what happened before your changes to what happens now - depending on the finding the change might be expected or not

**Comment by huoyaoyuan** [comment]
@krwq The method looks only used on XSLT parser.
After searching, it should be normal IEEE754 number, with dedicated representation of `INF`, `-INF` and `NaN`. I think there shouldn't be corner cases specified to this.

**Review by krwq — APPROVED** [review]

**Pull request merged**